pkg/source/kafka_scram_client.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package source
+
+import (
+	"crypto/sha256"
+	"crypto/sha512"
+	"hash"
+
+	"github.com/xdg/scram"
+)
+
+// SHA256 hash generator function for SCRAM conversation
+var SHA256 scram.HashGeneratorFcn = func() hash.Hash { return sha256.New() }
+
+// SHA512 hash generator function for SCRAM conversation
+var SHA512 scram.HashGeneratorFcn = func() hash.Hash { return sha512.New() }
+
+// XDGSCRAMClient struct to perform SCRAM conversation
+type XDGSCRAMClient struct {
+	*scram.Client
+	*scram.ClientConversation
+	scram.HashGeneratorFcn
+}
+
+// Begin starts SCRAM conversation
+func (x *XDGSCRAMClient) Begin(userName, password, authzID string) (err error) {
+	x.Client, err = x.HashGeneratorFcn.NewClient(userName, password, authzID)
+	if err != nil {
+		return err
+	}
+	x.ClientConversation = x.Client.NewConversation()
+	return nil
+}
+
+// Step performs step in SCRAM conversation
+func (x *XDGSCRAMClient) Step(challenge string) (response string, err error) {
+	response, err = x.ClientConversation.Step(challenge)
+	return
+}
+
+// Done completes SCRAM conversation
+func (x *XDGSCRAMClient) Done() bool {
+	return x.ClientConversation.Done()
+}
```

Comment:
What is this and why we need it? (sorry, i know the question might be stupid)

Reply:
That's what you must implement w/ doing the "none plain" SASL, w/ Sarama. They do here: https://github.com/Shopify/sarama/blob/c53ebd09c72909bb3b7b0bc8d9bfe8fa6b7bde39/examples/sasl_scram_client/scram_client.go

I guess... we could factor that into theirs? If they want, at a later point? I see others do _same_ ... as well ... :shrug:

Reply:
Ah, then i guess it's fine

Reply:
FWIW: https://github.com/Shopify/sarama/issues/1837